pkgs/development/compilers/llvm/common/default.nix

```diff
@@ -436,10 +436,16 @@ let
               && !stdenv.targetPlatform.isFreeBSD
               && stdenv.targetPlatform.useLLVM or false
             ) "-lunwind"
-            ++ lib.optional stdenv.targetPlatform.isWasm "-fno-exceptions";
+            ++ lib.optional stdenv.targetPlatform.isWasm "-fno-exceptions"
+            ++ lib.optional (lib.versionAtLeast metadata.release_version "17") [
+              "-Wno-unused-function"
+              "-Wno-unused-private-field"
+              "-Wno-implicit-function-declaration"
+            ];
           nixSupport.cc-ldflags = lib.optionals (
             !stdenv.targetPlatform.isWasm && !stdenv.targetPlatform.isFreeBSD
-          ) [ "-L${targetLlvmLibraries.libunwind}/lib" ];
+          ) [ "-L${targetLlvmLibraries.libunwind}/lib" ]
+            ++ lib.optional (lib.versionAtLeast metadata.release_version "17") "--undefined-version";
```

Comment:
i see how this creates less work in nixpkgs for projects that are only built with gcc normally, though i don't think it is helpful if one wants to use clang in external projects and we are disabling various warnings by default.

Reply:
These are actually marked as errors by default by Clang, we disable them so we can actually build packages and not have a million PR's.

Reply:
yeah, this doesn't seem like a proper solution to me. people use clang because if the additional checks. removing `Wno-implicit-function-declaration` makes it easier to write bad C code. seeing this. i would not use clang from nixpkgs for my project.

`unused-function` and `unused-private-field` are more benign but also things i would want enabled in my projects.

i think just doing `-Wno-xxx` completely disables the warning as well, which is not good. `-Wno-error=xxxx` would be better so at least the warnings are still there.

idk -- for the hardening flags one can turn them on or off by setting various attributes or env vars. maybe some attributes / env to be able to disable / enable a class of errors. perhaps just a list of packages that need the errors disabled and then can disable the errors just for them.

disabling them by default cripples clang. what do other distributions do?

Reply:
I'm not sure what other distros do, are there any distros built entirely using LLVM/clang?

Reply:
> These are actually marked as errors by default by Clang, we disable them so we can actually build packages and not have a million PR's.

this is incorrect. unused-function and unused-private-field are not errors and the warnings appear to be off by default. i had to add the flag `-Wunused` to enable them, and then they are only warnings and one needs to add `-Werror` for errors.

so it seems to me that this is the problem with packages specifying  `-Werror`.

this change is misrepresented and lacking in context, not to mention that gcc with -Wunused will also warn about unused functions.

source:
```c++
#include <stdio.h>

class Foo {

public:
    Foo() {};

    void hello(void) { printf("hello %p\n", this); };

private:
    int v;
};

static void print(const char *msg) { printf("%s\n", msg); }

int main(void)
{
    Foo f;
    f.hello();
    return 0;
}
```

i wrote a basic c and c++ tester and compiled with clang 18 and gcc14. no warnings, no errors.

please do not commit this change.

Reply:
Alright, thank you for the explanation. I removed the cflags changes but kept the ldflags change

Reply:
This really shouldn't be disabled. Misconfigured version scripts can be considered a bug in application code, not just a bad practice https://reviews.llvm.org/D135402. Leaving symbols undefined causes crashes and undefined behavior if wrong/nonexistent symbols are called during runtime. In Gentoo there were ~10 cases where the warning flip broke:

https://github.com/llvm/llvm-project/issues/61208#issuecomment-1456740248

Such issues have long been fixed so there shouldn't be a reason to disable this anymore, at least not by default. For instance, in zlib-ng this was fixed over a year ago https://github.com/zlib-ng/zlib-ng/issues/1427

Reply:
> Misconfigured version scripts can be considered a bug in application code,

How do version script affect the runtime?

> Leaving symbols undefined causes crashes and undefined behavior if wrong/nonexistent symbols are called during runtime

Wouldn't that affect the behavior under GCC as well?

> Such issues have long been fixed so there shouldn't be a reason to disable this anymore, at least not by default.

So we should use it only under packages which break? What should we do when packages refuse to fix it upstream?